onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -391,17 +387,34 @@ class PlannerImpl {
         if (!pnode->Name().empty()) errormsg << " (node " << pnode->Name() << ")";
         return Status(ONNXRUNTIME, FAIL, errormsg.str());
       }
-
       auto exec_provider = execution_providers_.Get(*pnode);
       if (exec_provider == nullptr) {
         return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Can not find the execution provider ",
                                pnode->GetExecutionProviderType());
       }
 
+      auto inputs = pnode->InputDefs();
+      auto num_inputs = inputs.size();
+      for (size_t i = 0; i < num_inputs; ++i) {
+        if (inputs[i]->Exists()) {
+          UseCount(inputs[i]->Name())++;
+          if (graph_inputs.end() != graph_inputs.find(inputs[i]->Name())) {
+            // If it's a graph input, set its plan.
```

Comment:
It would be easier to do quite a few things (especially in the optimizers) if we had edges from the graph inputs (and initializers) to nodes, and edges from nodes to graph outputs. Is that something we should consider (separate PR). It would avoid this sort of 'iterate all nodes to find the small number that involve graph inputs/outputs' logic.